### ISSUE TITLE
 Computational efficiency for very long or high-definition stage time series

### ISSUE THREAD

**Issue by benRenard** [BaRatin-tools/BaRatinAGE#40]
Val suggested two improvements to decrease computing times and storage requirements:
- Lossless compression of the stage time series: if 3 values are aligned, remove the middle one. Val tried it for a 25-year, 10-min stage series, and went down from 870 309 to 256 302 points - with no loss!
- Transformation into discharge by value rather than by time step. Because of rounding, the stage time series above containing 870 309 time steps only contains 4004 distinct stage values! It would therefore be much more efficient to transform only these 4004 values, and then put them back in the correct order to rebuild the time series. It's actually a bit more complicated than that because of stage errors, but it's certainly feasible to round the noisy stage values using the same resolution as the original stage data.

More generally, this raises the question to offer some pre-processing tools for input datasets, in relation with issue #26. Subsampling tools would be particularly valuable - but is it BaRatinAGE's job, or should rather be done by some external app?

**Comment by JeromeLeCoz** [comment]
Just an idea (maybe not relevant): would splitting long series into sub series help manage memory issues? If the series is too long, BaRatinAGE could apply the discharge series computation to several sub files, successively, and concantenate the results at the end.

Also, I saw a user case with many identical values (zeros actually) repeated in the same stage series. It may be helpful to detect identical (or similar within some tolerance interval?) values in a stage series and avoid repeating the discharge computations, to save time?

**Comment by benRenard** [comment]
> Just an idea (maybe not relevant): would splitting long series into sub series help manage memory issues? If the series is too long, BaRatinAGE could apply the discharge series computation to several sub files, successively, and concantenate the results at the end.

Yes this trick could be implemented in BaRatinAGE to avoid virtual memory issues (but do they happen frequently?). However it wouldn't change anything to storage issues (the size of the concatenated spag file will just be the sum of the smaller spag files) or  computing times (I think there is an outer loop over time steps, so treating 10^6 times steps or 100 batches of 10^4 time steps is the same).

> Also, I saw a user case with many identical values (zeros actually) repeated in the same stage series. It may be helpful to detect identical (or similar within some tolerance interval?) values in a stage series and avoid repeating the discharge computations, to save time?

This is similar to the fact that aligned values could be removed without information loss. But the more I think about it the more I think this is NOT a job for BaRatinAGE. This would make much more sense to have this (or other compression utilities) implemented in Octave (or as an external, stand-alone tool)

**Comment by JeromeLeCoz** [comment]
Yes, Octave already does what is needed (data compression in stage series). However not all BaRatinAGE know and/or have access to Octave... I even don't think that Octave is easy to find and download out of the French hydrological services. However it is open source and already does exactly what we need here. Would it be possible to embed (or recycle) the proper Octave function in BaRatinAGE? Or re-direct the user to the Octvae download (but I suspect this is not convenient...). This is not BaRatinAGE's job indeed, but without this being done, BaRatinAGE does not work...

**Comment by benRenard** [comment]
> Would it be possible to embed (or recycle) the proper Octave function in BaRatinAGE?

I have never seen Octave code so I have no idea how easy it would be to isolate the code that implements data compression utilities. But rather than directly re-implement this into BaRatinAGE, I think it would be better to re-implement this as a stand-alone utility (probably a command-line exe). Then in a second step BaRatinAGE can call this utility to perform data compression within BaRatinAGE. The advantage is that any other software/code can also call the same utility, without re-implementing it in Java/R/Python/etc. depending on the language of the calling software/code. The same reasoning applies to other type of tools, e.g. tools for aggregating or re-gridding time series. All these tools do not need to know anything about all the objects and concepts manipulated by BaRatinAGE, so I don't see any added value in integrating them deeply into BaRatinAGE's code. But it's just my feeling, maybe @IvanHeriver will think differently ? 

Maybe the first step would be to get in touch with Octave's developers to ask them how easy it would be to isolate and make stand-alone specific sets of utilities ?

**Comment by JeromeLeCoz** [comment]
Yes, I can provide you with the contacts of the Octave developers (ITLink) who are now developing Jacinthe and I-OLIV.

**Comment by IvanHeriver** [comment]
Are you sure Octave is open source? On the GDH website (https://gdh-hydrometrie.org/boite-a-outils/logiciels/tous-les-logiciels/) access to the software is provided upon asking it from the hydro 3 team by email (or the ITLink team as  suggested by Jérôme). 

Ben's suggestion sounds good to me. It follows the general philosophy followed so far i.e. having CLI tools (BaM, distribution) that can be called by anyone (BaRatinAGE, RBaM).

**Comment by JeromeLeCoz** [comment]
<img width="653" height="299" alt="Image" src="https://github.com/user-attachments/assets/226ccbe0-f7d4-4093-a61b-85a897b6e7a0" />

**Comment by JeromeLeCoz** [comment]
It comes with a Cecill license, however it seems it is released on demand.
